File: algorithms/2_medium/F/03862_find_the_smallest_balanced_index/solution.c
int smallestBalancedIndex(const int* nums, const int len) {
    int64_t sum_pref = 0;
    for (int i = 0; i < len; i += 1) {
        assert(nums[i] > 0);
        sum_pref += nums[i];
    }

    int64_t mul_suff = 1;
    for (int i = len - 1; i >= 0; i -= 1) {
        sum_pref -= nums[i];

        if (sum_pref == mul_suff) {
            return i;
        } else if (mul_suff > sum_pref) {
            return -1;
        }

        if (__builtin_mul_overflow(mul_suff, (int64_t)nums[i], &mul_suff)) {
            return -1;
        }
    }

    return -1;
}

File: algorithms/2_medium/F/03862_find_the_smallest_balanced_index/solution.rs
impl Solution {
    pub fn smallest_balanced_index(nums: Vec<i32>) -> i32 {
        let len = nums.len();
        debug_assert!(nums.iter().all(|x| *x > 0));

        let mut sum_pref: i64 = nums.iter().map(|n| *n as i64).sum();
        let mut mul_suff = 1 as i64;
        let mut overflow = false;
        for i in (1..len).rev() {
            sum_pref -= nums[i] as i64;

            if sum_pref == mul_suff {
                return i.try_into().unwrap();
            } else if mul_suff > sum_pref {
                break;
            }

            (mul_suff, overflow) = mul_suff.overflowing_mul(nums[i] as i64);
            if overflow {
                return -1;
            }
        }

        -1
    }
}
